`connected_to` shouldn't be called on abstract class that not established connection

alpaca-tc opened the pull request:
### Summary

Fixed: https://github.com/rails/rails/issues/40559#issuecomment-752056106

When abstract class hasn't own connections, calling `AbstractClass.connection` returns parent class's connection. We call `AbstractClass.connection.preventing_writes?` expecting abstract class's state to be returned, but actually it is parent's one.

I think that it isn't expected behavior so I prevents call `connected_to` on abstract
class that not established the connection.

~In addition, `connected_to_many` didn't check for passed classes, so now it checks for them as well as `connected_to`.~  This change is a not same concern, so I'll separate PR.

### Other Information

[bug report template gist](https://gist.github.com/alpaca-tc/f4335960b312e026808712a7f43a7947)

eileencodes commented:
I'm on vacation until Monday the 4th and will review this then.

eileencodes commented on activerecord/test/cases/base_test.rb:
```diff
@@ -1638,17 +1643,17 @@ def test_protected_environments_are_stored_as_an_array_of_string
   end
 
-  test "cannot call connected_to with role and shard on non-abstract classes" do
+  test "cannot call connected_to with role and shard on abstract class that not established the connection" do
```
Instead of changing all the tests can you just add a test for the new change?

eileencodes commented on activerecord/lib/active_record/connection_handling.rb:
```diff
@@ -377,5 +379,11 @@ def swap_connection_handler(handler, &blk) # :nodoc:
+
+      def check_own_connection_established(klass)
+        unless klass.name == klass.connection_specification_name
+          raise NotImplementedError, "calling `connected_to` is only allowed on abstract class that established the connection."
```
I don't think this gives the user enough information. There's a big difference between calling `connected_to` on a non-abstract class and calling it on the connection that doesn't have an established connection. It's also confusing that `connected_to` and `connected_to_many` handle this error differently. It should be handled the same. Can you update this to handle those cases separately and reinstate the previous error message when it's not an abstract class? 

Also there's a `the` missing in `is only allowed on abstract class` should be `is only allowed on the abstract class`.

alpaca-tc committed:
`connected_to` shouldn't be called on the abstract class that not established connection

Fixed: https://github.com/rails/rails/issues/40559#issuecomment-752056106

When abstract class hasn't own connections, calling `AbstractClass.connection`
returns parent class's connection. We call `AbstractClass.connection.preventing_writes?`
expecting abstract class's state to be returned, but actually it is parent's one.

I think that it isn't expected behavior so I prevents call `connected_to` on the abstract
class that not established the connection.

alpaca-tc commented:
@eileencodes I have been fixed 😄 
NOTE: I removed the commit for `connected_to_many` because this change is a not same concern, so I'll separate PR.

eileencodes commented:
I merged this manually on the command line in 6f0f9e086a7b589d54a9bb18573cff0dbbcecb16 because there were conflicts to fix and the tests didn't work after some other changes.

eileencodes commented:
Backported to 6-1-stable in bbb3a55e488b18072f616293dd494eac0c39adb4
